Remove pandas.core.index.datetimelike from MyPy Blacklist

makbigc opened the pull request:
- [x] closes #25882 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

makbigc committed:
Add type annotation and remove blacklist item

WillAyd commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -69,7 +70,6 @@ class DatetimeIndexOpsMixin(ExtensionOpsMixin):
     _resolution = cache_readonly(DatetimeLikeArrayMixin._resolution.fget)
```
So it looks like these `cache_readonly` calls are what's tripping up mypy. @jbrockmendel might be able to offer some insight here

jbrockmendel replied:
Is the problem with isort or mypy?

jbrockmendel replied:
If its mypy then would it be that it wants something like `# type: cache_readonly`?

WillAyd replied:
It's a mypy failure. Not actually with the cache_readonly as much as accessing the `.fget` attribute of the properties.

What's the intention of this code? Is passing fget from the property to cache_readonly faster than just accessing the property?

jbrockmendel replied:
> What's the intention of this code? Is passing fget from the property to cache_readonly faster than just accessing the property?

`cache_readonly behaves like `property`, expecting to get a unary method as its argument.  Does it even work if you don't make the `.fget` explicit?

makbigc replied:
This failure is probably due to mypy. [https://github.com/python/mypy/issues/6185](url)

Deleting `.fget` run into other errors.

WillAyd reviewed (changes requested):
Thanks for the PR! Looks like there's an isort failure in CI

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -134,7 +134,7 @@ def values(self) -> np.ndarray:
         return self._data._data
 
-    @property
+    @property  # type: ignore
```
why is this needed?

makbigc committed:
isort datetimelike.py

makbigc committed:
Add comment for the need of type: ignore

WillAyd commented:
Can you merge master?

vaibhavhrt commented:
@makbigc some of the errors in this file has been solved by #26404 
You can go ahead and fix(or ignore) the remaining.

WillAyd commented:
Can you merge master? Think this will be the last remaining module from the blacklist

jreback commented:
can you merge master

makbigc committed:
merge

makbigc committed:
Add ignore comment

makbigc commented:
mypy has the following error for the annotated value function: 

`pandas/core/indexes/datetimelike.py:139: error: "None" has no attribute "_data"`

I add annotation ` # type: DatetimeArray ` to `_data`. But mypy still has the same error.
https://github.com/pandas-dev/pandas/blob/8154efb0c1a64295cf54e00025b4ab09bcd02752/pandas/core/indexes/datetimes.py#L327

Do you have any better way?

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -130,11 +134,11 @@ def _ndarray_values(self):
     # Abstract data attributes
 
     @property
```
why this change?

makbigc replied:
mypy has the following error for the annotated value function: 

`pandas/core/indexes/datetimelike.py:139: error: "None" has no attribute "_data"`

I add annotation ` # type: DatetimeArray ` to `_data`. But mypy still has the same error.
https://github.com/pandas-dev/pandas/blob/8154efb0c1a64295cf54e00025b4ab09bcd02752/pandas/core/indexes/datetimes.py#L327

I guess that mypy mixed up the instance attribute `_data` with the class attributes `_data` which is set `None` in the beginning of DatetimeIndexOpsMixin.

jreback replied:
cc @WillAyd

WillAyd replied:
Sorry missed this ping. This could be a similar conversation to https://github.com/pandas-dev/pandas/pull/26518/files#r287573810

WillAyd replied:
I think a better solution is to provide the appropriate type to `_data`. Can you create a `DatetimeLikeArray` in pandas._typing which should be something like `DatetimeLikeArray = TypeVar('DatetimeLikeArray', DatetimeArray, PeriodArray, TimedeltaArray)` and assign that type to `_data` in the class? I think that should resolve.

cc @jbrockmendel in case he has other insights on the types here

WillAyd commented:
@makbigc can you merge master and address last comment? This should close out the reference issue (updated OP to reflect)

makbigc committed:
Add annotation DatetimeLikeArray to _data

makbigc committed:
resolve conflict

makbigc commented:
I can't figure out why DatetimeArray, PeriodArray and TimedeltaArray cannot be imported inside _typing.py

`from pandas.core.arrays import DatetimeArray, PeriodArray, TimedeltaArray`

jreback commented on pandas/_typing.py:
```diff
@@ -11,12 +11,17 @@
     ABCExtensionArray, ABCIndexClass, ABCSeries, ABCSparseSeries)
 
+from pandas.core.arrays import DatetimeArray, PeriodArray, TimedeltaArray
+
```
don’t try to import these use a quoted version

makbigc replied:
The following is the import error in the failed tests:
`ImportError: cannot import name '_INT64_DTYPE'`

The name of the module _typing.py is started with underscore. In PEP 8 ([https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles](url) ), it states

> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.

Should we remove underscore from the name _typing? Or something else?

jreback replied:
you can’t import these names as they would cause circular imports 
and don’t need to just use ‘DatetimeArray’ and such

makbigc replied:
`_INT64_DTYPE` is just an alias, i.e.,  `_INT64_DTYPE = np.dtype(np.int64)`. The import of numpy and the definition of `_INT64_DTYPE` are in the same file `pandas/core/dtype/common.py`.

I did a little experiment. Open a file called 'typing.py' (without underscore prefix) in the same directory as '_typing.py'. One can import `DatetimeArray` in `typing.py`.

Importing `PeriodArray` and `TImedeltaArray` also run into the same `ImportError`. Without them, one cannot define the mypy object `TypeVar`.
`DatetimeLikeArray = TypeVar('DatetimeLikeArray', DatetimeArray, PeriodArray, TimedeltaArray)`.

makbigc replied:
@WillAyd Would you tell me your opinion? Should we change the filename of _typing.py?

makbigc committed:
Import EA directly

WillAyd commented:
No I don’t think this needs to be renamed. I don’t have a problem using the TYPE_CHECKING constant in that module alone if it resolves circular imports. 

Sent from my iPhone

> On Jun 18, 2019, at 10:23 AM, Mak Sze Chun <notifications@github.com> wrote:
> 
> @makbigc commented on this pull request.
> 
> In pandas/_typing.py:
> 
> > @@ -11,12 +11,17 @@
>  from pandas.core.dtypes.generic import (
>      ABCExtensionArray, ABCIndexClass, ABCSeries, ABCSparseSeries)
>  
> +from pandas.core.arrays import DatetimeArray, PeriodArray, TimedeltaArray
> +
> @WillAyd Would you tell me your opinion? Should we change the filename of _typing.py?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

makbigc commented:
Yes, the problem is due to the circular import.
`common` -> `_typing` -> `datetimelikey` -> `common`
I can't resolve this circular import and datetimelike.py is the last exception in mypy.ini. So I close this PR for others picking it up.

WillAyd commented:
Doesn’t using the TYPE_CHECKING constant from the stdlib typing module in pandas._typing resolve the circular import?

I think this is really close!

makbigc commented:
@WillAyd Thanks for your reply. I have added the following in `_typing.py`
```
if TYPE_CHECKING:
    from pandas.core.arrays.datetimes import DatetimeArray
    from pandas.core.arrays.period import PeriodArray
    from pandas.core.arrays.timedeltas import TimedeltaArray
    DatetimeLikeArray = TypeVar('DatetimeLikeArray', DatetimeArray,
                                PeriodArray, TimedeltaArray)
```
and added the following to datetimelike.py
```
if TYPE_CHECKING:
    from pandas._typing import DatetimeLikeArray
```
The circular import is solved. But mypy cannot find the DatetimeLikeArray!
`pandas/core/indexes/datetimelike.py:67: error: Invalid type "pandas.DatetimeLikeArray"`

WillAyd committed:
Removed DatetimeLikeArray TypeVar for now

WillAyd committed:
Removed DatetimeLikeArray annotations

WillAyd committed:
isort fixup

WillAyd committed:
Update _typing.py

jreback commented:
lgtm. @WillAyd merge when ready.

WillAyd commented:
Thanks @makbigc !
